R/filter_counts.R

```diff
@@ -35,18 +37,39 @@ filter_counts <- function(sce, cr_like = TRUE, fdr_cutoff = 0.01, seed = NULL, .
     stop("cr_like must be set as TRUE or FALSE")
   }
 
+  if(!is.numeric(umi_cutoff) & umi_cutoff >= 0){
+    stop("umi_cutoff must be an integer greater than or equal to 0")
+  }
+
   # calculate probability of being an empty droplet
   if(cr_like){
-    empty_df <- DropletUtils::emptyDropsCellRanger(m = round(counts(sce)), ...)
+    empty_df <- tryCatch(
+      DropletUtils::emptyDropsCellRanger(m = round(counts(sce))),
+      error = function(x){NULL}
+    )
+    metadata(sce)$filtering_method <- "emptyDropsCellRanger"
   } else {
-    empty_df <- DropletUtils::emptyDrops(counts(sce), ...)
+    empty_df <- tryCatch(
+      DropletUtils::emptyDrops(counts(sce), ...),
+      error = function(x){NULL}
+    )
+    metadata(sce)$filtering_method <- "emptyDrops"
   }
 
-  if(any(empty_df$FDR > fdr_cutoff & empty_df$Limited, na.rm = TRUE)){
-    warning(glue::glue("`niters` may be set too low for emptyDrops filtering.",
-                    " Current value is {empty_df@metadata$niters}."))
+  # if emptyDrops doesn't fail, filter by FDR cutoff
+  if(!is.null(empty_df)){
+    if(any(empty_df$FDR > fdr_cutoff & empty_df$Limited, na.rm = TRUE)){
+      warning(glue::glue("`niters` may be set too low for emptyDrops filtering.",
+                         " Current value is {empty_df@metadata$niters}."))
+    }
+    cells <- rownames(empty_df)[which(empty_df$FDR <= fdr_cutoff)]
+  } else {
+    # if emptyDrops fails, filter using a hard UMI threshold
+    cells <- sce$sum > umi_cutoff
```

Comment:
I think `>=` aligns better with the statement in the docs ("minimum UMI count"), and also makes more intuitive sense to me.

```suggestion
    cells <- sce$sum >= umi_cutoff
```